ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java

```diff
@@ -33,21 +32,73 @@
 import org.apache.tuweni.units.bigints.UInt256;
 
 /** A World State backed first by trie log layer and then by another world state. */
-public class BonsaiLayeredWorldState implements BonsaiWorldState, WorldState {
+public class BonsaiLayeredWorldState implements BonsaiWorldView, WorldState {
+
+  private final BonsaiWorldView parent;
+  protected final long height;
+  protected final TrieLogLayer trieLog;
 
-  private final BonsaiWorldState parent;
-  private final TrieLogLayer trieLog;
-  private final Map<Address, Account> cachedAccounts = new HashMap<>();
+  private final Hash worldStateRootHash;
 
-  public BonsaiLayeredWorldState(final BonsaiWorldState parent, final TrieLogLayer trieLog) {
-    checkArgument(trieLog.isFrozen(), "TrieLogs must be frozen to be used as world state.");
+  BonsaiLayeredWorldState(
+      final BonsaiWorldView parent,
+      final long height,
+      final Hash worldStateRootHash,
+      final TrieLogLayer trieLog) {
     this.parent = parent;
+    this.height = height;
+    this.worldStateRootHash = worldStateRootHash;
     this.trieLog = trieLog;
   }
 
+  public BonsaiWorldView getParent() {
+    return parent;
+  }
+
+  public TrieLogLayer getTrieLog() {
+    return trieLog;
+  }
+
+  public long getHeight() {
+    return height;
+  }
+
   @Override
   public Bytes getCode(final Address address) {
-    return trieLog.getCode(address).orElseGet(() -> parent.getCode(address));
+    // this must be iterative and lambda light because the stack may blow up
+    // mainly because we don't have tail calls.
+    BonsaiLayeredWorldState currentLayer = this;
+    while (currentLayer != null) {
+      final Optional<Bytes> maybeCode = currentLayer.trieLog.getCode(address);
+      if (maybeCode.isPresent()) {
+        return maybeCode.get();
+      }
+      if (currentLayer.parent == null) {
+        currentLayer = null;
+      } else if (currentLayer.parent instanceof BonsaiLayeredWorldState) {
+        currentLayer = (BonsaiLayeredWorldState) currentLayer.parent;
+      } else {
+        return currentLayer.parent.getCode(address);
+      }
+    }
+    return null;
```

Comment:
Should we not return an OptionalEmpty instead ? Or did you do that to not impact Forest ?

Reply:
done.  It doesn't escape the bonsai classes.